File: common/src/main/java/net/shiroha233/roadweaver/structures/types/RoadsideStructure.java
/* 文件职责：定义路边结构的编解码协议与放置行为。 */
package net.shiroha233.roadweaver.structures.types;

import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Vec3i;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.ChunkPos;
import net.minecraft.world.level.StructureManager;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.chunk.ChunkGenerator;
import net.minecraft.world.level.levelgen.Heightmap;
import net.minecraft.world.level.levelgen.structure.BoundingBox;
import net.minecraft.world.level.levelgen.structure.Structure;
import net.minecraft.world.level.levelgen.structure.StructureStart;
import net.minecraft.world.level.levelgen.structure.StructureType;
import net.minecraft.world.level.levelgen.structure.pieces.PiecesContainer;
import net.minecraft.world.level.levelgen.structure.pieces.StructurePiecesBuilder;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
import net.minecraft.world.level.levelgen.heightproviders.ConstantHeight;
import net.minecraft.world.level.levelgen.heightproviders.HeightProvider;
import net.shiroha233.roadweaver.structures.data.LootConfig;
import net.shiroha233.roadweaver.structures.data.MobSpawnRule;
import net.shiroha233.roadweaver.structures.data.RoadsidePlacementRule;
import net.shiroha233.roadweaver.structures.data.StructureScale;
import net.shiroha233.roadweaver.structures.pieces.SimpleTemplatePiece;
import net.shiroha233.roadweaver.structures.registry.ModStructureTypes;

import java.util.List;
import java.util.Optional;

/**
 * 路边结构类型
 */
public class RoadsideStructure extends Structure {
    
    public static final MapCodec<RoadsideStructure> CODEC = RecordCodecBuilder.mapCodec(instance ->
        instance.group(
            settingsCodec(instance),
            ResourceLocation.CODEC.fieldOf("template").forGetter(s -> s.templateId),
            Vec3i.CODEC.optionalFieldOf("size_hint", new Vec3i(5, 5, 5)).forGetter(s -> s.sizeHint),
            Codec.INT.optionalFieldOf("weight", 10).forGetter(s -> s.weight),
            Codec.BOOL.optionalFieldOf("face_road", true).forGetter(s -> s.faceRoad),
            StructureScale.CODEC.optionalFieldOf("scale", StructureScale.SMALL).forGetter(s -> s.scale),
            RoadsidePlacementRule.CODEC.optionalFieldOf("placement_rule", RoadsidePlacementRule.UNIVERSAL).forGetter(s -> s.placementRule),
            MobSpawnRule.LIST_CODEC.optionalFieldOf("mob_spawns", List.of()).forGetter(s -> s.mobSpawns),
            LootConfig.LIST_CODEC.optionalFieldOf("loot_configs", List.of()).forGetter(s -> s.lootConfigs),
            HeightProvider.CODEC.optionalFieldOf("start_height", ConstantHeight.ZERO).forGetter(RoadsideStructure::startHeight),
            Heightmap.Types.CODEC.optionalFieldOf("project_start_to_heightmap", Heightmap.Types.WORLD_SURFACE_WG).forGetter(s -> s.projectStartToHeightmap),
            Codec.BOOL.optionalFieldOf("skip_terrain_check", false).forGetter(s -> s.skipTerrainCheck),
            Codec.BOOL.optionalFieldOf("use_accurate_height", false).forGetter(s -> s.useAccurateHeight)
        ).apply(instance, RoadsideStructure::new)
    );
    
    private final ResourceLocation templateId;
    private final Vec3i sizeHint;
    private final int weight;
    private final boolean faceRoad;
    private final StructureScale scale;
    private final RoadsidePlacementRule placementRule;
    private final List<MobSpawnRule> mobSpawns;
    private final List<LootConfig> lootConfigs;
    private final HeightProvider startHeight;
    private final Heightmap.Types projectStartToHeightmap;
    private final boolean skipTerrainCheck;
    private final boolean useAccurateHeight;
    
    public RoadsideStructure(
        StructureSettings settings,
        ResourceLocation templateId,
        Vec3i sizeHint,
        int weight,
        boolean faceRoad,
        StructureScale scale,
        RoadsidePlacementRule placementRule,
        List<MobSpawnRule> mobSpawns,
        List<LootConfig> lootConfigs,
        HeightProvider startHeight,
        Heightmap.Types projectStartToHeightmap,
        boolean skipTerrainCheck,
        boolean useAccurateHeight
    ) {
        super(settings);
        this.templateId = templateId;
        this.sizeHint = sizeHint;
        this.weight = weight;
        this.faceRoad = faceRoad;
        this.scale = scale;
        this.placementRule = placementRule;
        this.mobSpawns = mobSpawns;
        this.lootConfigs = lootConfigs;
        this.startHeight = startHeight;
        this.projectStartToHeightmap = projectStartToHeightmap;
        this.skipTerrainCheck = skipTerrainCheck;
        this.useAccurateHeight = useAccurateHeight;
    }
    
    @Override
    protected Optional<GenerationStub> findGenerationPoint(GenerationContext context) {
        return Optional.empty();
    }
    
    @Override
    public StructureType<?> type() {
        return ModStructureTypes.ROADSIDE;
    }
    
    public ResourceLocation templateId() {
        return templateId;
    }
    
    public Vec3i sizeHint() {
        return sizeHint;
    }
    
    public int weight() {
        return weight;
    }
    
    public boolean faceRoad() {
        return faceRoad;
    }
    
    public StructureScale scale() {
        return scale;
    }
    
    public RoadsidePlacementRule placementRule() {
        return placementRule;
    }
    
    public List<MobSpawnRule> mobSpawns() {
        return mobSpawns;
    }
    
    public List<LootConfig> lootConfigs() {
        return lootConfigs;
    }

    public HeightProvider startHeight() {
        return startHeight;
    }

    public Heightmap.Types projectStartToHeightmap() {
        return projectStartToHeightmap;
    }

    public boolean skipTerrainCheck() {
        return skipTerrainCheck;
    }

    public boolean useAccurateHeight() {
        return useAccurateHeight;
    }
    
    public StructureStart placeAt(WorldGenLevel level,
                                  StructureManager structureManager,
                                  StructureTemplateManager templateManager,
                                  ChunkGenerator generator,
                                  BlockPos pos,
                                  Rotation rotation,
                                  RandomSource random) {
        StructurePiecesBuilder builder = new StructurePiecesBuilder();
        
        SimpleTemplatePiece piece = new SimpleTemplatePiece(
            templateManager,
            templateId,
            pos,
            rotation,
            Mirror.NONE,
            mobSpawns,
            lootConfigs
        );
        builder.addPiece(piece);
        
        ChunkPos chunkPos = new ChunkPos(pos);
        PiecesContainer container = builder.build();
        StructureStart start = new StructureStart(this, chunkPos, 0, container);
        
        if (!start.isValid()) {
            return null;
        }
        
        BoundingBox boundingBox = start.getBoundingBox();
        start.placeInChunk(level, structureManager, generator, random, boundingBox, chunkPos);
        
        return start;
    }
    
    public void saveToChunk(WorldGenLevel level, 
                           StructureManager structureManager,
                           StructureStart start,
                           ChunkPos chunkPos) {
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/structures/types/RoadsideStructure.kt
package net.shiroha233.roadweaver.structures.types

import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.BlockPos
import net.minecraft.core.Vec3i
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.RandomSource
import net.minecraft.world.level.ChunkPos
import net.minecraft.world.level.StructureManager
import net.minecraft.world.level.WorldGenLevel
import net.minecraft.world.level.block.Mirror
import net.minecraft.world.level.block.Rotation
import net.minecraft.world.level.chunk.ChunkGenerator
import net.minecraft.world.level.levelgen.structure.Structure
import net.minecraft.world.level.levelgen.structure.StructureStart
import net.minecraft.world.level.levelgen.structure.StructureType
import net.minecraft.world.level.levelgen.structure.pieces.PiecesContainer
import net.minecraft.world.level.levelgen.structure.pieces.StructurePiecesBuilder
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager
import net.shiroha233.roadweaver.structures.data.LootConfig
import net.shiroha233.roadweaver.structures.data.MobSpawnRule
import net.shiroha233.roadweaver.structures.data.RoadsidePlacementRule
import net.shiroha233.roadweaver.structures.data.StructureScale
import net.shiroha233.roadweaver.structures.pieces.SimpleTemplatePiece
import net.shiroha233.roadweaver.structures.registry.ModStructureTypes
import java.util.Optional

/**
 * 路边结构类型
 *
 * 继承原版 Structure，用于数据驱动的路边结构定义。
 *
 * 特点：
 * - findGenerationPoint 返回 empty（不参与原版调度）
 * - 提供 placeAt 方法用于手动放置
 * - 支持保存到区块数据
 * - 完全通过 datapack JSON 定义
 */
@Suppress("unused")
class RoadsideStructure(
    settings: StructureSettings,
    private val templateId: ResourceLocation,
    private val sizeHint: Vec3i,
    private val weight: Int,
    private val faceRoad: Boolean,
    private val scale: StructureScale,
    private val placementRule: RoadsidePlacementRule,
    private val mobSpawns: List<MobSpawnRule>,
    private val lootConfigs: List<LootConfig>
) : Structure(settings) {
    /**
     * 原版调度入口 - 返回 empty，不参与自动调度
     */
    override fun findGenerationPoint(context: GenerationContext): Optional<GenerationStub> {
        return Optional.empty()
    }

    override fun type(): StructureType<*> {
        return ModStructureTypes.ROADSIDE
            ?: throw IllegalStateException("ModStructureTypes.ROADSIDE 未初始化：请确认平台层已完成结构类型注册")
    }

    // ==================== 属性访问器 ====================

    fun templateId(): ResourceLocation = templateId

    fun sizeHint(): Vec3i = sizeHint

    fun weight(): Int = weight

    fun faceRoad(): Boolean = faceRoad

    fun scale(): StructureScale = scale

    fun placementRule(): RoadsidePlacementRule = placementRule

    fun mobSpawns(): List<MobSpawnRule> = mobSpawns

    fun lootConfigs(): List<LootConfig> = lootConfigs

    // ==================== 手动放置方法 ====================

    /**
     * 在指定位置手动放置结构
     *
     * @return 如果成功放置则返回 StructureStart，否则返回 null
     */
    fun placeAt(
        level: WorldGenLevel,
        structureManager: StructureManager,
        templateManager: StructureTemplateManager,
        generator: ChunkGenerator,
        pos: BlockPos,
        rotation: Rotation,
        random: RandomSource
    ): StructureStart? {
        val builder = StructurePiecesBuilder()

        val piece = SimpleTemplatePiece(
            templateManager,
            templateId,
            pos,
            rotation,
            Mirror.NONE,
            mobSpawns,
            lootConfigs
        )
        builder.addPiece(piece)

        val chunkPos = ChunkPos(pos)
        val container: PiecesContainer = builder.build()
        val start = StructureStart(this, chunkPos, 0, container)

        if (!start.isValid) {
            return null
        }

        val boundingBox = start.boundingBox
        start.placeInChunk(level, structureManager, generator, random, boundingBox, chunkPos)

        return start
    }

    /**
     * 将 StructureStart 保存到区块数据
     */
    fun saveToChunk(
        level: WorldGenLevel,
        structureManager: StructureManager,
        start: StructureStart,
        chunkPos: ChunkPos
    ) {
        // 原版会自动处理保存，这里只是标记
        // StructureManager 会在区块保存时自动序列化 StructureStart
    }

    companion object {
        @JvmField
        val CODEC: MapCodec<RoadsideStructure> = RecordCodecBuilder.mapCodec { instance ->
            instance.group(
                settingsCodec(instance),
                ResourceLocation.CODEC.fieldOf("template").forGetter { it.templateId },
                Vec3i.CODEC.optionalFieldOf("size_hint", Vec3i(5, 5, 5)).forGetter { it.sizeHint },
                Codec.INT.optionalFieldOf("weight", 10).forGetter { it.weight },
                Codec.BOOL.optionalFieldOf("face_road", true).forGetter { it.faceRoad },
                StructureScale.CODEC.optionalFieldOf("scale", StructureScale.SMALL).forGetter { it.scale },
                RoadsidePlacementRule.CODEC.optionalFieldOf("placement_rule", RoadsidePlacementRule.UNIVERSAL)
                    .forGetter { it.placementRule },
                MobSpawnRule.LIST_CODEC.optionalFieldOf("mob_spawns", listOf()).forGetter { it.mobSpawns },
                LootConfig.LIST_CODEC.optionalFieldOf("loot_configs", listOf()).forGetter { it.lootConfigs }
            ).apply(instance, ::RoadsideStructure)
        }
    }
}
